### PR TITLE
Permissible random Failure

### PR DESCRIPTION
**Pull request by kedarbcs16**

### CODE REVIEW

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 1]
```diff
@@ -143,7 +143,8 @@ public int runTests() throws Exception {
     List<DrillTest> timeoutFailures = Lists.newArrayList();
     List<DrillTest> canceledTests = Lists.newArrayList();
     List<DrillTest> randomFailures = Lists.newArrayList();
-
+    List<DrillTest> tempTobeRunFailures = Lists.newArrayList();
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: change to failingTests

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 2]
```diff
@@ -289,20 +290,23 @@ public int runTests() throws Exception {
       if (cmdParam.trackMemory) {
     	  queryMemoryUsage();
       }
-
+      if(tempTobeRunFailures.size()>0){
+        tempTobeRunFailures.clear();
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: might not be necessary. Please check.

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 3]
```diff
@@ -289,20 +290,23 @@ public int runTests() throws Exception {
...
-	      randomFailures.add(test);
-	   }
+	  //List<DrillTest> tempTests = Lists.newArrayList();
+          //tempTests.add(test);
+          //LOG.info(DrillTestDefaults.LINE_BREAK);
+          //LOG.info("ISOLATING RANDOM FAILURES - Execution attempt 2 (of 2)");
+          //LOG.info(DrillTestDefaults.LINE_BREAK);
+          //executor.executeAll(tempTests);
+          //testStatus = tempTests.get(0).getTestStatus();
+          //if(testStatus==TestStatus.PASS){
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Remove all commented out lines

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 4]
```diff
@@ -328,6 +332,43 @@ public int runTests() throws Exception {
         }
       }
 
+      //TO check
+      if(cmdParam.threads >= tempTobeRunFailures.size()){
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: make the comparison the other way round to make it more clear. Also add a comment. The idea here is that we run this random failure detection logic only when the number of failures are low. We are using the concurrency level as the maximum number of failures under which the logic kicks in.

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 5]
```diff
@@ -328,6 +332,43 @@ public int runTests() throws Exception {
...
+        LOG.info(DrillTestDefaults.LINE_BREAK);
+        LOG.info("ISOLATING RANDOM FAILURES - Execution attempt 2 (of 2)");
+        LOG.info(DrillTestDefaults.LINE_BREAK);
+        executor.executeAll(tempTobeRunFailures);
+	for(DrillTest test : tempTobeRunFailures){
+	  TestStatus testStatus = test.getTestStatus();         
+          if(testStatus==TestStatus.PASS){
+            randomFailures.add(test);
+          }
+
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Remove new line

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 6]
```diff
@@ -328,6 +332,43 @@ public int runTests() throws Exception {
...
+          }
+	
+	}
+      }
+      
+      if(randomFailures.size()>0){
+        LOG.info("\nRandom Failures:\n");
+        for (DrillTest test : randomFailures) {
+          LOG.info("Query: " + test.getInputFile() + "\n" + test.getQuery());
+          LOG.info("Exception:\n", test.getException());
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Remove these logs.

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 7]
```diff
@@ -328,6 +332,43 @@ public int runTests() throws Exception {
...
+	  TestStatus testStatus = test.getTestStatus();         
+          if(testStatus==TestStatus.PASS){
+            randomFailures.add(test);
+          }
+	
+	}
+      }
+      
+      if(randomFailures.size()>0){
+        LOG.info("\nRandom Failures:\n");
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Remove the log

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 8]
```diff
@@ -368,14 +409,6 @@ public int runTests() throws Exception {
...
       }
 
-      if(randomFailures.size()>0){
-        LOG.info("\nRandom Failures:\n");
-        for (DrillTest test : randomFailures) {
-          LOG.info("Query: " + test.getInputFile() + "\n" + test.getQuery());
-          LOG.info("Exception:\n", test.getException());
-        }
-      }
-
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Keep this here.

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 9]
```diff
@@ -328,6 +332,43 @@ public int runTests() throws Exception {
...
+	    dataVerificationFailures.remove(test);
+	  }
+	  else if(planVerificationFailures.contains(test)){
+	    planVerificationFailures.remove(test);
+	  }
+	  else if(timeoutFailures.contains(test)){
+	    timeoutFailures.remove(test);
+	  }
+
+
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Remove new lines

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java (hunk 2 above)
Comment (kedarbcs16) [reply framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Checked abhishek ,I think this is necessary and in fact needs to be added to the other data structure as well (passing tests,execution failures etc),when running for 2 iterations they pile on the existing data structure and produce duplicates :- 

Plan Verification Failures:

/root/drillAutomation/drill-test-framework/framework/resources/Functional/filter/pushdown/plan/q4.sql
/root/drillAutomation/drill-test-framework/framework/resources/Functional/filter/pushdown/plan/q4.sql

Total Tests                  : 8397
Passing Tests                : 16756 --(the number is more than the total tests)
Failing Tests                : 38

> Execution Failures         : 36 --(this just doubled up)
> Data Verification Failures : 0
> Plan Verification Failures : 2 --(this just doubled up)
> Timeout Failures           : 0

Comment (kedarbcs16) [reply framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: should I make the changes for them as well ?

**Pull request closed without merging**